pkg/clustervalidator/clustervalidator.go

```diff
@@ -0,0 +1,84 @@
+package clustervalidator
+
+import (
+	"context"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/util/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/aws/eks-anywhere/pkg/cluster"
+	"github.com/aws/eks-anywhere/pkg/retrier"
+)
+
+// ClusterValidation defines a validation that can be registered to the ClusterValidator.
+type ClusterValidation = func(ctx context.Context, vc Config) error
+
+// RetriableClusterValidation represents a ClusterValidation which can have settings to be retried.
+type RetriableClusterValidation struct {
+	Validation ClusterValidation // the validation to run against the cluster.
+	Retrier    *retrier.Retrier  // the retrier to define how to re-attempt the validation.
+}
+
+// ClusterValidator is responsible for checking if a cluster is valid against the spec that is provided.
+type ClusterValidator struct {
+	Config      Config
+	validations []RetriableClusterValidation
+}
+
+// WithValidations registers multiple validations to the ClusterValidator that will be run when Validate is called.
+func (c *ClusterValidator) WithValidations(retrieableValidations ...RetriableClusterValidation) {
+	c.validations = append(c.validations, retrieableValidations...)
+}
+
+// Validate runs through the set registered validations and returns an error if any of them fail after a number of retries.
+func (c *ClusterValidator) Validate(ctx context.Context) error {
+	var wg sync.WaitGroup
+	wg.Add(len(c.validations))
+	errList := make([]error, 0)
+	for _, validation := range c.validations {
+		v := validation
+		go func() {
+			defer wg.Done()
+			err := c.runValidation(ctx, v)
+			if err != nil {
+				errList = append(errList, err)
+			}
+		}()
+	}
+	wg.Wait()
+	return errors.NewAggregate(errList)
+}
+
+func (c *ClusterValidator) runValidation(ctx context.Context, v RetriableClusterValidation) error {
```

Comment:
This is a little peculiar. I would think the `ClusterValidator` has a set of `ClusterValidation` funcs and those funcs are responsible for retrying if they need to.

We could add a method to `RetriableClusterValidation` that matches the `ClusterValidation` signature and wraps up the retry algorithm. That makes the `RetriableClusterValidation` a helper of sorts that can still be registered with `ClusterValidator`. Then make `ClusterValidator` only concerned with `ClusterValidation`s and drop the `runValidation` func creating a much clearer separation of concerns.

E.g.
```go
// Add a method complimenting the ClusterValidation 
func (RetriableClusterValidation) Validate(ctx context.Context, vc Config) error { ... }

// Change to only care about ClusterValidation funcs.
type ClusterValidator struct {
	Config      Config
	validations []ClusterValidation
}

// Register the retriable validation Validate method with the validator instead.
func doValidation() {  
  retriable := RetriableValidation{
    Retrier: NewRetrier(),
    Validation: func(ctx context.Context, vc Config) error {
      // do validation
    }
  }

  validator := NewClusterValidator()
  validator.WithValidations(retriable.Validate)
  
}
```

Reply:
Okay, interesting. This would definitely be preferred as well for registering validations that possible don't require retries